File: app/src/main/java/app/opass/ccip/fragment/MainFragment.java
package app.opass.ccip.fragment;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.design.widget.Snackbar;
import android.support.design.widget.TextInputEditText;
import android.support.design.widget.TextInputLayout;
import android.support.v4.app.Fragment;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AlertDialog;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;
import app.opass.ccip.R;
import app.opass.ccip.activity.CaptureActivity;
import app.opass.ccip.activity.MainActivity;
import app.opass.ccip.adapter.ScenarioAdapter;
import app.opass.ccip.model.Attendee;
import app.opass.ccip.network.CCIPClient;
import app.opass.ccip.util.PreferenceUtil;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.zxing.integration.android.IntentIntegrator;
import com.onesignal.OneSignal;
import org.json.JSONException;
import org.json.JSONObject;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainFragment extends Fragment {

    View noNetworkView;
    View notConfWifiView;
    View loginView;
    TextView loginTitle;
    RecyclerView scenarioView;
    SwipeRefreshLayout swipeRefreshLayout;
    private Activity mActivity;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        View view = inflater.inflate(R.layout.fragment_main, container, false);

        noNetworkView = view.findViewById(R.id.no_network);
        notConfWifiView = view.findViewById(R.id.not_conf_wifi);
        loginView = view.findViewById(R.id.login);
        loginTitle = view.findViewById(R.id.login_title);

        View enterTokenButton = view.findViewById(R.id.enter_token);

        enterTokenButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                final LayoutInflater inflater = mActivity.getLayoutInflater();
                final View dialogView = inflater.inflate(R.layout.dialog_enter_token, null);
                final TextInputLayout tokenInputLayout = dialogView.findViewById(R.id.token_input_layout);
                final TextInputEditText tokenInput = dialogView.findViewById(R.id.token_input);

                final AlertDialog dialog = new AlertDialog.Builder(mActivity)
                    .setView(dialogView)
                    .setTitle(R.string.enter_your_token)
                    .setPositiveButton(R.string.positive_button, null)
                    .setNegativeButton(R.string.negative_button, null)
                    .create();

                dialog.setOnShowListener(new DialogInterface.OnShowListener() {
                    @Override
                    public void onShow(DialogInterface dialogInterface) {
                        Button positiveButton = dialog.getButton(DialogInterface.BUTTON_POSITIVE);
                        Button negativeButton = dialog.getButton(DialogInterface.BUTTON_NEGATIVE);

                        positiveButton.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                String token = tokenInput.getText().toString().trim();

                                if (token.equals("")) {
                                    tokenInputLayout.setError(getString(R.string.token_required));
                                    return;
                                }
                                PreferenceUtil.setIsNewToken(mActivity, true);
                                PreferenceUtil.setToken(mActivity, token);
                                dialog.dismiss();
                                updateStatus();
                            }
                        });
                        negativeButton.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                dialog.dismiss();
                            }
                        });
                    }
                });

                dialog.show();
            }
        });

        loginTitle.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                IntentIntegrator integrator = new IntentIntegrator(mActivity);
                integrator.setDesiredBarcodeFormats(IntentIntegrator.QR_CODE);
                integrator.setPrompt(getString(R.string.scan_kktix_qrcode));
                integrator.setCameraId(0);
                integrator.setBeepEnabled(false);
                integrator.setBarcodeImageEnabled(false);
                integrator.setCaptureActivity(CaptureActivity.class);
                integrator.initiateScan();
            }
        });
        scenarioView = view.findViewById(R.id.scenarios);
        swipeRefreshLayout = view.findViewById(R.id.swipeContainer);

        mActivity = getActivity();
        scenarioView.setLayoutManager(new LinearLayoutManager(mActivity));
        scenarioView.setItemAnimator(new DefaultItemAnimator());

        if (mActivity.getIntent().getAction().equals(Intent.ACTION_VIEW)) {
            String token = mActivity.getIntent().getData().getQueryParameter("token");

            if (token != null) {
                PreferenceUtil.setIsNewToken(mActivity, true);
                PreferenceUtil.setToken(mActivity, token);
            }
        }

        if (PreferenceUtil.getToken(mActivity) == null) {
            loginView.setVisibility(View.VISIBLE);
        }

        swipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                updateStatus();
            }
        });

        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        updateStatus();
    }

    void updateStatus() {
        if (PreferenceUtil.getToken(mActivity) == null) {
            loginView.setVisibility(View.VISIBLE);
            return;
        }

        swipeRefreshLayout.post(new Runnable() {
            @Override
            public void run() {
                swipeRefreshLayout.setRefreshing(true);
            }
        });
        loginView.setVisibility(View.GONE);
        noNetworkView.setVisibility(View.GONE);
        notConfWifiView.setVisibility(View.GONE);

        Call<Attendee> attendee = CCIPClient.get().status(PreferenceUtil.getToken(mActivity));
        attendee.enqueue(new Callback<Attendee>() {
            @Override
            public void onResponse(Call<Attendee> call, Response<Attendee> response) {
                swipeRefreshLayout.setRefreshing(false);
                if (response.isSuccessful()) {
                    Attendee attendee = response.body();
                    JsonObject attr = attendee.getAttr().getAsJsonObject();

                    if (PreferenceUtil.getIsNewToken(mActivity)) {
                        PreferenceUtil.setIsNewToken(mActivity, false);

                        JSONObject tags = new JSONObject();
                        try {
                            tags.put("event_id", attendee.getEventId());
                            tags.put("token", attendee.getToken());
                            tags.put("type", attendee.getType());
                            OneSignal.sendTags(tags);
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }

                        new AlertDialog.Builder(mActivity)
                            .setMessage(mActivity.getString(R.string.hi)
                                + attendee.getUserId()
                                + mActivity.getString(R.string.login_success))
                            .setPositiveButton(android.R.string.ok, null)
                            .show();
                    }

                    JsonElement attrTitle = attr.get("title");
                    if (attrTitle != null) {
                        MainActivity.setUserTitle(attrTitle.getAsString());
                    }
                    MainActivity.setUserId(attendee.getUserId());

                    scenarioView.setAdapter(new ScenarioAdapter(mActivity, attendee.getScenarios()));
                } else if (response.code() == 403) {
                    swipeRefreshLayout.setRefreshing(false);
                    notConfWifiView.setVisibility(View.VISIBLE);
                    notConfWifiView.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            swipeRefreshLayout.setRefreshing(true);
                            notConfWifiView.setVisibility(View.GONE);
                            updateStatus();
                        }
                    });
                } else {
                    Snackbar.make(getView(), getString(R.string.invalid_token), Snackbar.LENGTH_LONG).show();
                    PreferenceUtil.setToken(mActivity, null);
                    loginView.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onFailure(Call<Attendee> call, Throwable t) {
                swipeRefreshLayout.setRefreshing(false);
                noNetworkView.setVisibility(View.VISIBLE);
                noNetworkView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        swipeRefreshLayout.setRefreshing(true);
                        noNetworkView.setVisibility(View.GONE);
                        updateStatus();
                    }
                });
            }
        });
    }
}

File: app/src/main/java/app/opass/ccip/fragment/MainFragment.kt
package app.opass.ccip.fragment

import android.content.Intent
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import app.opass.ccip.R
import app.opass.ccip.activity.AuthActivity
import app.opass.ccip.activity.CountdownActivity
import app.opass.ccip.activity.MainActivity
import app.opass.ccip.adapter.ScenarioAdapter
import app.opass.ccip.extension.asyncExecute
import app.opass.ccip.model.Scenario
import app.opass.ccip.network.CCIPClient
import app.opass.ccip.network.ErrorUtil
import app.opass.ccip.util.JsonUtil
import app.opass.ccip.util.PreferenceUtil
import com.google.android.material.snackbar.Snackbar
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import kotlin.coroutines.CoroutineContext

class MainFragment : Fragment(), CoroutineScope {
    companion object {
        private const val EXTRA_URL = "EXTRA_URL"
        fun newInstance(url: String): MainFragment = MainFragment().apply {
            arguments = Bundle().apply { putString(EXTRA_URL, url) }
        }
    }

    private lateinit var noNetworkView: View
    private lateinit var notConfWifiView: View
    private lateinit var loginView: View
    private lateinit var scenarioView: RecyclerView
    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var mActivity: MainActivity
    private lateinit var mJob: Job
    override val coroutineContext: CoroutineContext
        get() = mJob + Dispatchers.Main
    private lateinit var mAdapter: ScenarioAdapter

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        super.onCreate(savedInstanceState)
        val view = inflater.inflate(R.layout.fragment_main, container, false)

        mActivity = requireActivity() as MainActivity
        noNetworkView = view.findViewById(R.id.no_network)
        notConfWifiView = view.findViewById(R.id.not_conf_wifi)
        loginView = view.findViewById(R.id.login_view)
        mJob = Job()

        noNetworkView.setOnClickListener {
            noNetworkView.visibility = View.GONE
            updateStatus()
        }

        view.findViewById<View>(R.id.login_button).setOnClickListener {
            mActivity.startActivity(Intent(mActivity, AuthActivity::class.java))
        }

        scenarioView = view.findViewById(R.id.scenarios)
        swipeRefreshLayout = view.findViewById(R.id.swipeContainer)

        scenarioView.layoutManager = LinearLayoutManager(mActivity)
        scenarioView.itemAnimator = DefaultItemAnimator()

        if (PreferenceUtil.getToken(mActivity) == null) {
            loginView.visibility = View.VISIBLE
            swipeRefreshLayout.visibility = View.GONE
        }

        swipeRefreshLayout.setOnRefreshListener { updateStatus() }

        return view
    }

    override fun onResume() {
        super.onResume()
        updateStatus()
    }

    override fun onDestroy() {
        super.onDestroy()
        mJob.cancel()
    }

    private fun updateStatus() {
        val token = PreferenceUtil.getToken(mActivity)
        if (token == null) {
            loginView.visibility = View.VISIBLE
            swipeRefreshLayout.visibility = View.GONE
            return
        }

        swipeRefreshLayout.visibility = View.VISIBLE
        swipeRefreshLayout.post { swipeRefreshLayout.isRefreshing = true }
        loginView.visibility = View.GONE
        noNetworkView.visibility = View.GONE
        notConfWifiView.visibility = View.GONE

        launch {
            try {
                val response = CCIPClient.get().status(token).asyncExecute()

                when {
                    response.isSuccessful -> {
                        val attendee = response.body()
                        val attr = attendee!!.attr.asJsonObject

                        attr.get("title")?.let {
                            mActivity.setUserTitle(it.asString)
                        }
                        mActivity.setUserId(attendee.userId)

                        mAdapter = ScenarioAdapter(mActivity, attendee.scenarios) {
                            val isUsed = it.used != null
                            val hasCountdown = it.countdown > 0

                            if (isUsed && hasCountdown) {
                                startCountdownActivity(it)
                                return@ScenarioAdapter
                            }

                            if (hasCountdown) {
                                showConfirmDialog(it)
                            } else {
                                useScenario(it)
                            }
                        }
                        scenarioView.adapter = mAdapter
                    }
                    response.code() == 403 -> {
                        notConfWifiView.visibility = View.VISIBLE
                        notConfWifiView.setOnClickListener {
                            swipeRefreshLayout.isRefreshing = true
                            notConfWifiView.visibility = View.GONE
                            updateStatus()
                        }
                    }
                    else -> {
                        Snackbar.make(view!!, getString(R.string.invalid_token), Snackbar.LENGTH_LONG).show()
                        PreferenceUtil.setToken(mActivity, null)
                        PreferenceUtil.setRole(mActivity, null)
                        loginView.visibility = View.VISIBLE
                    }
                }
            } catch (t: Throwable) {
                noNetworkView.visibility = View.VISIBLE
            } finally {
                swipeRefreshLayout.isRefreshing = false
            }
        }
    }

    private fun useScenario(scenario: Scenario) {
        val token = PreferenceUtil.getToken(mActivity)

        launch {
            try {
                val response = CCIPClient.get().use(scenario.id, token).asyncExecute()
                when {
                    response.isSuccessful -> {
                        val attendee = response.body()
                        mAdapter.setItems(attendee!!.scenarios)

                        if (scenario.countdown > 0) {
                            startCountdownActivity(scenario)
                        }
                    }
                    response.code() == 400 -> {
                        val (message) = ErrorUtil.parseError(response)
                        Toast.makeText(mActivity, message, Toast.LENGTH_LONG).show()
                    }
                    response.code() == 403 -> {
                        AlertDialog.Builder(mActivity)
                            .setTitle(R.string.connect_to_conference_wifi)
                            .setPositiveButton(android.R.string.ok, null)
                            .show()
                    }
                    else -> Toast.makeText(mActivity, "Unexpected response", Toast.LENGTH_LONG).show()
                }
            } catch (t: Throwable) {
                Toast.makeText(mActivity, "Use req fail, " + t.message, Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun showConfirmDialog(scenario: Scenario) {
        AlertDialog.Builder(mActivity)
            .setTitle(R.string.confirm_dialog_title)
            .setPositiveButton(R.string.positive_button) { dialogInterface, i -> useScenario(scenario) }
            .setNegativeButton(R.string.negative_button, null)
            .show()
    }

    private fun startCountdownActivity(scenario: Scenario) {
        val intent = Intent()
        intent.setClass(mActivity, CountdownActivity::class.java)
        intent.putExtra(CountdownActivity.INTENT_EXTRA_SCENARIO, JsonUtil.toJson(scenario))
        mActivity.startActivity(intent)
    }
}
